### DISCUSSION TITLE
Support relations with the Repeater field

### DISCUSSION THREAD

**Discussion by Flynsarmy** [wintercms/winter#504] [Ideas]
`Relation` formwidget currently only supports JSON fields. I'd like it to also support model relations.

I made an example `RelatedRepeater` formwidget by extending `Repeater` to demonstrate how this might work. Code available here:
https://github.com/Flynsarmy/wn-relatedrepeater-plugin

The code is ugly as sin but it appears to work perfectly for hasMany relations and could be merged back into Repeater when done. I haven't tested other relation types.

Would love some testing for any edge cases, other relation types etc. I'm happy to keep working on this.

**Comment by bennothommo** [comment]
@Flynsarmy great work on this. My only notes would be:

- This explicitly would require the related model to support sorting, so a check would need to be done to ensure that the related model is using the `Sortable` trait.
- I imagine that this could *only* work for `hasMany` relations, as the morph relations would possibly require an extra step in selecting the target morph model. `hasOne` and `belongsTo` can already be handled by other fields, and `hasManyThrough` might not be easily established since there's an intermediate model.
- How would repeater limits be handled with a relation? (Or more to the point, what would occur if a limit is established on a repeater but that limit is breached by the number of related records? Would the other records be dropped?)
- Following on from the above, would removing the item from the repeater delete the relation?

**Reply by Flynsarmy** [reply]
Thanks Ben this was exactly what I needed!

> a check would need to be done to ensure that the related model is using the Sortable trait.

Added. When using a `RelationRepeater`, the formwidget will be sortable when the relation class uses the `Sortable` trait, and isn't sortable when the relation doesn't include the trait.

> I imagine that this could only work for hasMany relations

I've locked the widget to only work as a RelationRepeater when the relation is a `hasMany`. I'll still explore the other relation types but for the time being we can just start with this.

> what would occur if a limit is established on a repeater but that limit is breached by the number of related records

This is already covered by existing checks. A popup appears stating *option_values only allows up to 2 items, 3 were provided*. The admin will need to remove the excess records before continuing.

> would removing the item from the repeater delete the relation

It currently orphans the record in the database. I think it would be more appropriate to delete the record though.

My repo has been updated with the latest version of the formwidget.

**Comment by Flynsarmy** [comment]
Currently waiting on https://github.com/wintercms/winter/pull/506 before this can continue

**Reply by bennothommo** [reply]
Merged it @Flynsarmy

**Reply by Flynsarmy** [reply]
My repo has been updated to use the new sorting functionality. If I've addressed all your concerns above, I'll start a WIP PR so we can clean up the code. Otherwise let me know if there's anything you'd like changed.

**Reply by bennothommo** [reply]
@Flynsarmy happy for you to proceed with the PR.
